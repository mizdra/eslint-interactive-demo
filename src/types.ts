import { Icon } from './lib/icon';

export type FormData = {
  query: string;
};

export type CursorPosition = {
  styleTop: number;
  styleLeft: number;
};

/**
 * `presetIcons` オプションの要素の型
 * */
export type PresetIconsItem =
  | Icon
  | Icon[]
  | Promise<Icon | Icon[]>
  | (() => PresetIconsItem[])
  | (() => Promise<PresetIconsItem[]>);
