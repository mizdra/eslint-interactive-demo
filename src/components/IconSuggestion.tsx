import { FunctionComponent } from 'preact';

export const IconSuggestion: FunctionComponent = () => {
  return <div>text</div>;
};
