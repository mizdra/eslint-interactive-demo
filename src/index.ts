export { registerIconSuggestion } from './lib/register';
export { Icon } from './lib/icon';
export { fetchMemberPageIcons, fetchRelatedPageIconsByHashTag } from './lib/preset-icon';
