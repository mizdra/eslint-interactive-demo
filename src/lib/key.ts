/** IME による変換中かどうか */
export function isComposing(e: KeyboardEvent): boolean {
  return e.isComposing || (e.key === 'Enter' && e.which === 229);
}
