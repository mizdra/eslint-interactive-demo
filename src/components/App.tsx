import { FunctionComponent } from 'preact';
import { useCallback, useState } from 'preact/hooks';
import { forwardPartialFuzzyMatcher } from '..';
import { useDocumentEventListener } from '../hooks/useDocumentEventListener';
import { useScrapbox } from '../hooks/useScrapbox';
import { uniqueIcons } from '../lib/collection';
import { Icon } from '../lib/icon';
import { isComposing } from '../lib/key';
import { calcCursorPosition, insertText, scanEmbeddedIcons } from '../lib/scrapbox';
import { CursorPosition, Matcher } from '../types';
import { SearchablePopupMenu } from './SearchablePopupMenu';

const DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN = (e: KeyboardEvent) => {
  return e.key === 'l' && e.ctrlKey && !e.shiftKey && !e.altKey && !e.metaKey;
};

const DEFAULT_IS_INSERT_QUERY_KEY_DOWN = (e: KeyboardEvent) => {
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && e.altKey && !e.metaKey) return true;
  if (e.key === 'Enter' && !e.ctrlKey && !e.shiftKey && !e.altKey && e.metaKey) return true;
  return false;
};

export type AppProps = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
  isInsertQueryKeyDown?: (e: KeyboardEvent) => boolean;
  presetIcons?: Icon[];
  defaultSuggestPresetIcons?: boolean;
  matcher?: Matcher;
};

export const App: FunctionComponent<AppProps> = ({
  isSuggestionOpenKeyDown = DEFAULT_IS_SUGGESTION_OPEN_KEY_DOWN,
  isSuggestionCloseKeyDown,
  isInsertQueryKeyDown = DEFAULT_IS_INSERT_QUERY_KEY_DOWN,
  presetIcons = [],
  defaultSuggestPresetIcons = true,
  matcher = forwardPartialFuzzyMatcher,
}) => {
  const { textInput, cursor, editor, layout, projectName } = useScrapbox();

  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [embeddedIcons, setEmbeddedIcons] = useState<Icon[]>([]);
  const [suggestPresetIcons, setSuggestPresetIcons] = useState(defaultSuggestPresetIcons);
  const composedMatcher = useCallback(
    (query: string) => {
      const composedIcons = uniqueIcons(suggestPresetIcons ? [...embeddedIcons, ...presetIcons] : embeddedIcons);
      return matcher({ query, composedIcons, presetIcons, embeddedIcons });
    },
    [embeddedIcons, matcher, presetIcons, suggestPresetIcons],
  );
  const [query, setQuery] = useState('');

  const handleSelect = useCallback(
    (icon: Icon) => {
      setOpen(false);
      insertText(textInput, icon.getNotation(projectName));
    },
    [projectName, textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleSuggestionOpenKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      e.preventDefault();
      e.stopPropagation();

      if (!open) {
        // ポップアップが閉じていたら開く
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanEmbeddedIcons` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        // 画像化されたらエディタを走査してアイコンを収集
        const newEmbeddedIcons = scanEmbeddedIcons(projectName, editor);

        setEmbeddedIcons(newEmbeddedIcons);
        setOpen(true);
        setSuggestPresetIcons(defaultSuggestPresetIcons);
      } else {
        // ポップアップが開いていたら、preset icon の表示・非表示をトグルする
        setSuggestPresetIcons((suggestPresetIcons) => !suggestPresetIcons);
      }
    },
    [cursor, defaultSuggestPresetIcons, editor, open, layout, projectName, textInput],
  );

  const handleInsertQueryKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (layout !== 'page') return; // エディタのあるページ以外ではキー入力を無視する
      if (!open) return; // ポップアップが閉じていたら無視する
      e.preventDefault();
      e.stopPropagation();
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [layout, open, textInput, query],
  );

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (isComposing(e)) return; // IMEによる変換中は何もしない
      if (isSuggestionOpenKeyDown(e)) {
        handleSuggestionOpenKeyDown(e);
      } else if (isInsertQueryKeyDown(e)) {
        handleInsertQueryKeyDown(e);
      }
    },
    [isSuggestionOpenKeyDown, isInsertQueryKeyDown, handleSuggestionOpenKeyDown, handleInsertQueryKeyDown],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SearchablePopupMenu
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      cursorPosition={cursorPosition}
      matcher={composedMatcher}
      onSelect={handleSelect}
      onClose={handleClose}
      onInputQuery={setQuery}
      isSuggestionCloseKeyDown={isSuggestionCloseKeyDown}
    />
  );
};
