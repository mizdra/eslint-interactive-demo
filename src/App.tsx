import { FunctionComponent } from 'preact';
import { useCallback, useMemo, useState } from 'preact/hooks';
import { SuggestionBox, Item } from './components/SuggestionBox';
import { useDocumentEventListener } from './hooks/useDocumentEventListener';
import { uniqBy } from './lib/collection';
import { calcCursorPosition, insertText, scanIconsFromEditor } from './lib/scrapbox';
import { CursorPosition, Icon } from './types';

const cursor = document.querySelector<HTMLElement>('.cursor')!;

function uniqueIcons(icons: Icon[]): Icon[] {
  return uniqBy(icons, (icon) => icon.pagePath);
}

function generateItem(icon: Icon): Item {
  return {
    element: (
      <span>
        <img
          alt={icon.imgAlt}
          title={icon.imgTitle}
          style="width: 1.3em; height: 1.3em; object-fit: contain;"
          src={icon.imgSrc}
        />
        {' ' + icon.pagePath}
      </span>
    ),
    searchableText: icon.pagePath,
  };
}

type AppProps = {
  isSuggestionOpenKeyDown: (e: KeyboardEvent) => boolean;
  presetIcons: Icon[];
  editor: HTMLElement;
  textInput: HTMLTextAreaElement;
};

export const App: FunctionComponent<AppProps> = ({ isSuggestionOpenKeyDown, presetIcons, editor, textInput }) => {
  const [open, setOpen] = useState(false);
  const [cursorPosition, setCursorPosition] = useState<CursorPosition>({ styleTop: 0, styleLeft: 0 });
  const [suggestibleIcons, setSuggestibleIcons] = useState<Icon[]>([]);
  const suggestBoxItems = useMemo(() => suggestibleIcons.map(generateItem), [suggestibleIcons]);
  const [presetAppended, setPresetAppended] = useState(false);

  const handleSelect = useCallback(
    (_item: Item, index: number) => {
      setOpen(false);
      insertText(textInput, suggestibleIcons[index].notation);
    },
    [suggestibleIcons, textInput],
  );

  const handleSelectNonexistent = useCallback(
    (query: string) => {
      setOpen(false);
      insertText(textInput, `[${query}.icon]`);
    },
    [textInput],
  );

  const handleClose = useCallback(() => {
    setOpen(false);
    textInput.focus();
  }, [textInput]);

  const handleKeydown = useCallback(
    (e: KeyboardEvent) => {
      if (!isSuggestionOpenKeyDown(e)) return;
      e.preventDefault();
      e.stopPropagation();

      if (open && !presetAppended) {
        setSuggestibleIcons(uniqueIcons([...suggestibleIcons, ...presetIcons]));
        setPresetAppended(true);
      } else {
        const icons = scanIconsFromEditor(scrapbox.Project.name, editor);
        setCursorPosition(calcCursorPosition(cursor));

        // NOTE: ある行にフォーカスがあると、行全体がテキスト化されてしまい、`scanIconsFromEditor` で
        // アイコンを取得することができなくなってしまう。そのため、予めフォーカスを外し、フォーカスのあった
        // 行のアイコン記法が画像化されるようにしておく。
        textInput.blur();
        setSuggestibleIcons(uniqueIcons(icons));
        setOpen(true);
        setPresetAppended(false);
      }
    },
    [editor, isSuggestionOpenKeyDown, open, presetAppended, presetIcons, suggestibleIcons, textInput],
  );
  useDocumentEventListener('keydown', handleKeydown, { capture: true });

  return (
    <SuggestionBox
      open={open}
      emptyMessage="キーワードにマッチするアイコンがありません"
      items={suggestBoxItems}
      cursorPosition={cursorPosition}
      onSelect={handleSelect}
      onSelectNonexistent={handleSelectNonexistent}
      onClose={handleClose}
    />
  );
};
