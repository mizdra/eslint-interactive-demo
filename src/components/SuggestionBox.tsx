import { ComponentChild } from 'preact';
import { useCallback, useEffect, useMemo, useState } from 'preact/hooks';
import { JSXInternal } from 'preact/src/jsx';
import { forwardPartialFuzzyMatcher } from '../lib/matcher';
import { CursorPosition, Matcher } from '../types';
import { PopupMenu } from './PopupMenu';
import { QueryInput } from './SuggestionBox/QueryInput';

export type Item<T> = {
  key: JSXInternal.IntrinsicAttributes['key'];
  element: ComponentChild;
  searchableText: string;
  value: T;
};

export type SuggestionBoxProps<T> = {
  open: boolean;
  emptyMessage?: string;
  items: Item<T>[];
  cursorPosition: CursorPosition;
  matcher?: Matcher<T>;
  onSelect?: (item: Item<T>, query: string) => void;
  onSelectNonexistent?: (query: string) => void;
  onClose?: (query: string) => void;
  onInputQuery?: (query: string) => void;
  isSuggestionCloseKeyDown?: (e: KeyboardEvent) => boolean;
};

export function SuggestionBox<T>({
  open,
  emptyMessage,
  items,
  cursorPosition,
  matcher = forwardPartialFuzzyMatcher,
  onSelect,
  onSelectNonexistent,
  onClose,
  onInputQuery,
  isSuggestionCloseKeyDown,
}: SuggestionBoxProps<T>) {
  const [query, setQuery] = useState('');
  const matchedItems = useMemo(() => matcher(query, items), [items, matcher, query]);
  const matchedItemsForPopupMenu = useMemo(
    () =>
      matchedItems.map((item) => ({
        key: item.key,
        element: item.element,
      })),
    [matchedItems],
  );

  useEffect(() => {
    if (open === false) setQuery('');
  }, [open]);

  const handleSelect = useCallback(
    (_item: ComponentChild, index: number) => {
      onSelect?.(matchedItems[index], query);
    },
    [matchedItems, onSelect, query],
  );
  const handleSelectNonexistent = useCallback(() => {
    onSelectNonexistent?.(query);
  }, [onSelectNonexistent, query]);
  const handleClose = useCallback(() => {
    onClose?.(query);
  }, [onClose, query]);
  const handleInputQuery = useCallback(
    (query: string) => {
      setQuery(query);
      onInputQuery?.(query);
    },
    [onInputQuery],
  );

  return (
    <div>
      <PopupMenu
        open={open}
        emptyMessage={emptyMessage}
        items={matchedItemsForPopupMenu}
        cursorPosition={cursorPosition}
        onSelect={handleSelect}
        onSelectNonexistent={handleSelectNonexistent}
        onClose={handleClose}
        isPopupCloseKeyDown={isSuggestionCloseKeyDown}
      />
      {open && (
        <QueryInput
          defaultQuery={query}
          cursorPosition={cursorPosition}
          onInput={handleInputQuery}
          onBlur={handleClose}
        />
      )}
    </div>
  );
}
