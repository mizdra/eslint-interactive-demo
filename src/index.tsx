import { render } from 'preact';
import { App } from './App';
import { pagePathToIcon } from './lib/icon';
import { getEditor } from './lib/scrapbox';

type Options = {
  isSuggestionOpenKeyDown?: (e: KeyboardEvent) => boolean;
  // `string` is `pagePath`
  presetIcons?: string[];
  editor?: HTMLElement;
  textInput?: HTMLTextAreaElement;
  cursor?: HTMLElement;
  scrapbox?: Scrapbox;
};

export function registerIconSuggestion(options?: Options) {
  // 直接 editor に mount すると scrapbox 側の react renderer と干渉して壊れるので、
  // editor 内に差し込んだ container に mount する
  const container = document.createElement('div');
  const editor = options?.editor ?? getEditor();
  editor.appendChild(container);

  const presetIcons = options?.presetIcons?.map((pagePath) => pagePathToIcon(scrapbox.Project.name, pagePath));

  render(
    <App
      isSuggestionOpenKeyDown={options?.isSuggestionOpenKeyDown}
      presetIcons={presetIcons}
      editor={editor}
      textInput={options?.textInput}
      cursor={options?.cursor}
      scrapbox={options?.scrapbox}
    />,
    container,
  );
}
