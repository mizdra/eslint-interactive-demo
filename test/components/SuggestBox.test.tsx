import { act, fireEvent, render } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { datatype } from 'faker';
import { matchItems, SuggestionBox } from '../../src/components/SuggestionBox';
import { CursorPosition } from '../../src/types';
import { keydownEnterEvent, keydownEscapeEvent } from '../helpers/key';

// ダミーの props
const cursorPosition: CursorPosition = { styleTop: 0, styleLeft: 0 };
const items = [
  { element: <span key="1">a</span>, searchableText: 'a', value: 'a' },
  { element: <span key="2">ab</span>, searchableText: 'ab', value: 'ab' },
  { element: <span key="3">abc</span>, searchableText: 'abc', value: 'abc' },
  { element: <span key="4">z</span>, searchableText: 'z', value: 'z' },
];
const props = { cursorPosition, items };

describe('matchItems', () => {
  test('query に部分一致する items のみが返る', () => {
    expect(
      matchItems('foo', [
        // マッチする
        { element: '', searchableText: 'foo', value: '' },
        { element: '', searchableText: 'foo bar', value: '' },
        { element: '', searchableText: 'foo bar baz', value: '' },
        { element: '', searchableText: 'bar foo baz', value: '' },
        { element: '', searchableText: 'bar baz foo', value: '' },
        { element: '', searchableText: 'foo foo foo', value: '' },
        // マッチしない
        { element: '', searchableText: 'bar', value: '' },
        { element: '', searchableText: 'fo bar', value: '' },
        { element: '', searchableText: 'fo o bar', value: '' },
      ]),
    ).toStrictEqual([
      { element: '', searchableText: 'foo', value: '' },
      { element: '', searchableText: 'foo bar', value: '' },
      { element: '', searchableText: 'foo bar baz', value: '' },
      { element: '', searchableText: 'bar foo baz', value: '' },
      { element: '', searchableText: 'bar baz foo', value: '' },
      { element: '', searchableText: 'foo foo foo', value: '' },
    ]);
  });
  test('マッチは capital-insensitive', () => {
    expect(
      matchItems('foo', [
        { element: '', searchableText: 'Foo', value: '' },
        { element: '', searchableText: 'FOO', value: '' },
        { element: '', searchableText: 'fOo', value: '' },
      ]),
    ).toStrictEqual([
      { element: '', searchableText: 'Foo', value: '' },
      { element: '', searchableText: 'FOO', value: '' },
      { element: '', searchableText: 'fOo', value: '' },
    ]);
    expect(
      matchItems('FOO', [
        { element: '', searchableText: 'Foo', value: '' },
        { element: '', searchableText: 'FOO', value: '' },
        { element: '', searchableText: 'fOo', value: '' },
      ]),
    ).toStrictEqual([
      { element: '', searchableText: 'Foo', value: '' },
      { element: '', searchableText: 'FOO', value: '' },
      { element: '', searchableText: 'fOo', value: '' },
    ]);
  });
});

describe('SuggestionBox', () => {
  describe('open === false の時', () => {
    test('ポップアップも QueryInput も表示されない', () => {
      const { asFragment, queryByTestId } = render(<SuggestionBox open={false} {...props} />);
      expect(queryByTestId('popup-menu')).toBeNull();
      expect(queryByTestId('query-input')).toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('open === true の時', () => {
    test('ポップアップと QueryInput が表示される', () => {
      const { asFragment, queryByTestId } = render(<SuggestionBox open {...props} />);
      expect(queryByTestId('popup-menu')).not.toBeNull();
      expect(queryByTestId('query-input')).not.toBeNull();
      expect(asFragment()).toMatchSnapshot();
    });
    test('Esc 押下で onClose が呼び出される', async () => {
      const onClose = jest.fn();
      render(<SuggestionBox open {...props} onClose={onClose} />);
      expect(onClose).toBeCalledTimes(0);
      await act(() => {
        fireEvent(document, keydownEscapeEvent);
      });
      expect(onClose).toBeCalledTimes(1);
    });
    describe('ポップアップに表示されるアイテムが空の時', () => {
      test('emptyMessage でアイテムが空の時のメッセージを変更できる', () => {
        const emptyMessage = datatype.string();
        const { getByText } = render(<SuggestionBox open {...props} items={[]} emptyMessage={emptyMessage} />);
        expect(getByText(emptyMessage)).toBeInTheDocument();
      });
      test('Enter 押下で onSelectNonexistent が呼び出される', async () => {
        const onSelectNonexistent = jest.fn();
        render(<SuggestionBox open {...props} items={[]} onSelectNonexistent={onSelectNonexistent} />);
        expect(onSelectNonexistent).toBeCalledTimes(0);
        await act(() => {
          fireEvent(document, keydownEnterEvent);
        });
        expect(onSelectNonexistent).toBeCalledTimes(1);
      });
    });
    describe('ポップアップに表示されるアイテムが空でない時', () => {
      test('QueryInput に文字を入力するとアイテムがフィルタされる', () => {
        const { getByTestId } = render(<SuggestionBox open {...props} />);
        const buttonContainer = getByTestId('button-container');
        const queryInput = getByTestId('query-input');

        expect(buttonContainer.childElementCount).toEqual(4);
        userEvent.type(queryInput, 'a');
        expect(buttonContainer.childElementCount).toEqual(3);
        userEvent.type(queryInput, 'b');
        expect(buttonContainer.childElementCount).toEqual(2);
        userEvent.type(queryInput, 'c');
        expect(buttonContainer.childElementCount).toEqual(1);
        userEvent.type(queryInput, 'd');
        expect(buttonContainer.childElementCount).toEqual(0);
      });
      test('Enter 押下で onSelect が呼び出される', async () => {
        const onSelect = jest.fn();
        render(<SuggestionBox open {...props} onSelect={onSelect} />);
        expect(onSelect).toBeCalledTimes(0);
        await act(() => {
          fireEvent(document, keydownEnterEvent);
        });
        expect(onSelect).toBeCalledTimes(1);
      });
    });
  });
  test('open === true になった時に、 QueryInput に入力された文字がリセットされる', () => {
    const { rerender, getByTestId } = render(<SuggestionBox open {...props} />);

    userEvent.type(getByTestId('query-input'), 'a');
    expect(getByTestId('query-input')).toHaveValue('a');

    rerender(<SuggestionBox open={false} {...props} />);
    rerender(<SuggestionBox open {...props} />);

    expect(getByTestId('query-input')).toHaveValue('');
  });
});
