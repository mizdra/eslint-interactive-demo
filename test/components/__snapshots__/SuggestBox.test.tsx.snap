// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SuggestionBox open === false の時 ポップアップも QueryInput も表示されない 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`SuggestionBox open === true の時 ポップアップと QueryInput が表示される 1`] = `
<DocumentFragment>
  <div>
    <div
      class="popup-menu"
      data-testid="popup-menu"
      style="top: 0px;"
    >
      <div
        class="button-container"
        data-testid="button-container"
        style="left: 0px; transform: translateX(-NaN%);"
      >
        <div
          class="button selected"
        >
          <span>
            a
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            ab
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            abc
          </span>
        </div>
        <div
          class="button"
        >
          <span>
            z
          </span>
        </div>
      </div>
      <div
        class="triangle"
        style="left: 0px;"
      />
    </div>
    <form
      class="form-inline"
    >
      <input
        class="form-control"
        data-testid="query-input"
        default="true"
        style="position: absolute; top: 0px; left: 0px; transform: translateX(-NaN%);"
      />
    </form>
  </div>
</DocumentFragment>
`;
